### ISSUE TITLE
The problem of include

### ISSUE THREAD

**Issue by dcy** [vim-erlang/vim-erlang-compiler#25]
`include("xxx.hrl")` will be `can't find include file "xxx.hrl"`
`include("include/xxx.hrl")` it works.

**Comment by dcy** [comment]
Can anyone help me with this?

**Comment by hcs42** [comment]
Could you provide some more information? Which build system do you use: nothing, rebar 2, rebar 3, Makefile? What is the current directory (`:pwd`) when you want to compile? Do you use `:make` to start the compilation?

When you do a :make, Vim prints a line like this:

    :!/home/.../vim-erlang-compiler/compiler/erlang_check.erl --outdir ebin ... myfile.erl 2>&1| tee /var/folders/r3/cj0sy26x2kg5bcb9g48btljh0000gn/T/vq4rwHZ/196

Could you run the same command from the terminal, and append a "-v" option as well for verbosity? E.g.

    /home/.../vim-erlang-compiler/compiler/erlang_check.erl --outdir ebin ... myfile.erl -v

(You need to be in a directory so that the relative path at the end will be correct. Or you can use an absolute file name and then the current directory won't matter.)

Could you attach the output as a txt file to this ticket?

**Comment by dcy** [comment]
Not use rebar3, just vim show the error.
![2016-06-03-172947_609x432_scrot](https://cloud.githubusercontent.com/assets/1137238/15774778/cee964e4-29b0-11e6-8186-03a51b3ca5de.png)

**Comment by dcy** [comment]
I use these Erlang plugins:
```
"NeoBundle 'jimenezrick/vimerl'
NeoBundle 'vim-erlang/vim-erlang-compiler'
NeoBundle 'vim-erlang/vim-erlang-skeletons'
NeoBundle 'vim-erlang/vim-erlang-runtime'
NeoBundle 'vim-erlang/vim-erlang-omnicomplete'
NeoBundle 'vim-erlang/vim-erlang-tags'
NeoBundle 'scrooloose/syntastic'
```

**Comment by hcs42** [comment]
Thanks for the further info. I managed to reproduce the problem and created a fix. Could you try it? It's on my include_fix branch: https://github.com/hcs42/vim-erlang-compiler/commits/include_fix

**Comment by dcy** [comment]
In my another computer, no the error.
I will try it in my company computer tomorrow.

**Comment by dcy** [comment]
I don't know why it will raise the error in my company computer, but it will not raise error in my home computer, the same vim config.
https://github.com/hcs42/vim-erlang-compiler/commits/include_fix will not raise any error, include syntax error.
Maybe you can test in your computer, just a xxx.hrl in include/, ```-include("xxx.hrl").``` in erl.

**Comment by hcs42** [comment]
I tested it on my computer, and the original branch produced the error you mentioned, while the include_fix branch didn't produce this error.

> https://github.com/hcs42/vim-erlang-compiler/commits/include_fix will not raise any error, include syntax error.

Do you mean that the include_fix branch works well on your company computer too?

**Comment by aboroska** [comment]
I can confirm that https://github.com/hcs42/vim-erlang-compiler/commits/include_fix fixes the problem. (I was able to reproduce the issue.) Thanks @hcs42 !

**Comment by dcy** [comment]
I download the zip file from https://github.com/hcs42/vim-erlang-compiler/commits/include_fix, then unzip it and change name "vim-erlang-compile", mv it into ```~/.vim/bundle/```.
It will not raise the error ```can't find include file "xxx.hrl"```, but the other syntax error can not raise too.
![2016-06-07-093727_349x464_scrot](https://cloud.githubusercontent.com/assets/1137238/15843281/88f3ba94-2c93-11e6-9120-32c2991d95b7.png)

**Comment by dcy** [comment]
Hi! This error will raise in my some applications, not every application.
![2016-06-07-202132_786x257_scrot](https://cloud.githubusercontent.com/assets/1137238/15857360/968560ea-2ced-11e6-9485-9e08e038f324.png)

**Comment by hcs42** [comment]
I fixed my fix. Could you try again? (I changed "file:dirname" to "filename:dirname".)

**Comment by dcy** [comment]
It works. Thanks. I don't know if i download the wrong branch code to test.
It works now.

**Comment by hcs42** [comment]
> I don't know if i download the wrong branch code to test.

No, my fix was faulty when your checked it first, sorry about that. Although I tried it out and so did @aboroska, but it's a mystery to me how it worked for both of us :)

Anyway, I merged the fix into the main branch (https://github.com/vim-erlang/vim-erlang-compiler/pull/27), and I'm closing this ticket.

Thanks for reporting the problem and for the screenshots.

**Comment by dcy** [comment]
@hcs42 src/temp/xxx.erl have the same error.

```include("xxx.hrl")``` will be can't find include file "xxx.hrl"
```include("include/xxx.hrl")``` it works.

**Comment by hcs42** [comment]
Is my understanding correct that it worked previously (when we closed this ticket)?

Could you test which commit in vim-erlang-compiler broke it again?

**Comment by dcy** [comment]
```src/xxx.erl``` works, ```src/temp/xxx.erl``` does not work.

**Comment by hcs42** [comment]
Thanks for reporting the problem, I managed to reproduce the error. I'm reopening this ticket.

**Comment by hcs42** [comment]
I have pushed a (not very elegant) fix to [my hcs42/vim-erlang-compiler/include_fix branch](https://github.com/hcs42/vim-erlang-compiler/commits/include_fix) – could you try it?

**Comment by dcy** [comment]
@hcs42 Maybe https://github.com/vim-erlang/vim-erlang-compiler/pull/39 is better

**Comment by vihu** [comment]
Another instance where I notice problems with `include` is when you have `hrl` files in the `src` directory of the application. For example to keep those `hrl` files private, we sometimes put those in the src directory.

I tried to address that here: https://github.com/vihu/vim-erlang-compiler/commit/8e74d6db662c0941ebc2985e6f51023552dbc24f but didn't have much luck.

I think it needs to build at the root level of the app and _then_ check the module or something along those lines?

**Comment by hcs42** [comment]
@vihu Am I understanding correctly that the issue is the following:

- the header file is e.g. `myapp/src/a.hrl`
- the source file is e.g. `myapp/src/mydir/b.erl`
- the source file says `-include("a.hrl")`

If yes:

- Doesn't the following work when added to the `check_module` function?

    ```
    {i, filename:join([AppRoot, "src"])}
    ```

- Also, does using `-include("../a.hrl")` solve the problem?

If not:

- could you describe what is the path of the problematic files and how they include each other?

**Comment by halturin** [comment]
i got this problem having following project structure

```
project/
  apps/
    myapp
      include
         myapp.hrl
      src
         myapp.erl
  rebar.config
```
myapp.erl has `-include_lib("myapp/include/myapp.hrl").`
shows me (if my current dir is the root of the project): can't find include lib "myapp/include/myapp.hrl"
but if I change my current dir to apps - everything is fine.

is that intended behavior?

**Comment by halturin** [comment]
just executed manually with -v option. got these lines
![image](https://user-images.githubusercontent.com/118860/102380087-cd47ab00-3fc7-11eb-929c-a8a3dc2a17e9.png)
why do you exclude the project dir? (`outdir` means excluding?)

**Comment by halturin** [comment]
just fixed using this workaround by adding these lines into my rebar.conf:

```
{erl_opts, [
	{i, "apps"}
]}.
```
but I believe it should be included by default, because `apps` directory is the standard naming for the multiapp project.

**Comment by hcs42** [comment]
@halturin The project structure you describe should work without changing `vim-erlang-compiler`.

However, this is true only after you issue `rebar3 compile` once (assuming you use `rebar3`). The reason is that `include_lib` finds Erlang applications based on their `*.app` files, and those files are generated only during `rebar compile`.

I added this information to the README now: https://github.com/vim-erlang/vim-erlang-compiler#i-get-compilation-error

Can you confirm whether `vim-erlang-compiler` compilation works after `rebar3 compile` without the workaround?

**Comment by hcs42** [comment]
> why do you exclude the project dir? (outdir means excluding?)

No, `--outdir DIR` means that `vim-erlang-compiler` will place the created BEAM file in `DIR`. See `erlang_check.erl --help` (or the [help text](https://github.com/vim-erlang/vim-erlang-compiler/blob/cd4f05a67016d0f0814333fa915ad975cb9fa12f/compiler/erlang_check.erl#L157) inside the `erlang_check.erl`).

I think the project dir is missing because the `*.app` files are not yet there.

**Comment by hcs42** [comment]
@halturin I assume that the solution I wrote in https://github.com/vim-erlang/vim-erlang-compiler/issues/25#issuecomment-748514612 solved the problem (otherwise let me know).

Also, I pushed a commit (5bbff6e029719c50678cf628ade00b32eb48641f) which I think solves @vihu's problem (see https://github.com/vim-erlang/vim-erlang-compiler/issues/25#issuecomment-566751347).

With that, I'm closing this ticket. If you experience another include problem, please open a new ticket. (Different problems manifest themselves through include problems, and it makes the tickets easier to maintain if different problems are in different tickets.)
